File: solutions/2026-03-20/cpp/2026-03-20-minimum-absolute-difference-in-sliding-submatrix.cpp
```cpp
/*
Problem Summary: Find the minimum absolute difference between any two distinct elements in each k x k submatrix of a given m x n grid.

Link: https://leetcode.com/problems/minimum-absolute-difference-in-sliding-submatrix/

Approach:
The problem requires us to iterate through all possible k x k submatrices and for each submatrix, find the minimum absolute difference.
A naive approach would be to extract each submatrix, flatten it into a 1D array, sort it, and then iterate through the sorted array to find the minimum difference. This would be too slow given the constraints.

A more efficient approach leverages the fact that the grid dimensions are small (up to 30x30). We can iterate through each possible top-left corner (i, j) of a k x k submatrix. For each submatrix:
1. Collect all k*k elements into a temporary vector.
2. Sort this vector.
3. Iterate through the sorted vector and calculate the minimum difference between adjacent elements. If all elements are the same, the difference is 0.

To optimize the collection of elements for overlapping submatrices, we could potentially use a sliding window approach with a data structure that efficiently maintains sorted elements and allows for updates (add/remove). However, given the small constraints, the simpler approach of re-collecting and re-sorting for each submatrix is feasible and easier to implement correctly.

The number of possible k x k submatrices is (m - k + 1) * (n - k + 1).
For each submatrix, we extract k*k elements, sort them (O(k*k log(k*k))), and then find the minimum difference (O(k*k)).

Time Complexity:
Let m be the number of rows and n be the number of columns in the grid.
The number of submatrices is (m - k + 1) * (n - k + 1).
For each submatrix, we extract k*k elements.
Sorting these k*k elements takes O(k*k * log(k*k)) time.
Finding the minimum difference after sorting takes O(k*k) time.
Thus, the total time complexity is approximately O((m - k + 1) * (n - k + 1) * k*k * log(k*k)).
Given m, n <= 30 and k <= min(m, n), k*k can be at most 30*30 = 900.
The dominant factor is the number of submatrices and the sorting within each.

Space Complexity:
We use a temporary vector to store the elements of a single k x k submatrix, which takes O(k*k) space.
The output array `ans` takes O((m - k + 1) * (n - k + 1)) space.
Therefore, the space complexity is O(k*k + (m - k + 1) * (n - k + 1)), which simplifies to O(m*n) in the worst case if k is small.
Since k*k <= m*n, it can be considered O(m*n).

Considerations for implementation:
- Iterate through the possible starting row `r` from 0 to `m - k`.
- Iterate through the possible starting column `c` from 0 to `n - k`.
- For each `(r, c)`, extract the submatrix elements from `grid[r...r+k-1][c...c+k-1]`.
- Store these elements in a `std::vector`.
- Sort the `std::vector`.
- Calculate the minimum difference between adjacent elements. Handle the case where all elements are the same.
- Store the result in `ans[r][c-start_col]`.

*/

#include <vector>
#include <algorithm>
#include <climits>

class Solution {
public:
    std::vector<std::vector<int>> minAbsoluteDifference(std::vector<std::vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();

        // The result matrix will have dimensions (m - k + 1) x (n - k + 1)
        std::vector<std::vector<int>> ans(m - k + 1, std::vector<int>(n - k + 1));

        // Iterate through all possible top-left corners of the k x k submatrices
        for (int r = 0; r <= m - k; ++r) {
            for (int c = 0; c <= n - k; ++c) {
                // This vector will store all elements of the current k x k submatrix
                std::vector<int> submatrix_elements;
                submatrix_elements.reserve(k * k); // Optimize by reserving space

                // Extract elements from the current submatrix
                for (int i = r; i < r + k; ++i) {
                    for (int j = c; j < c + k; ++j) {
                        submatrix_elements.push_back(grid[i][j]);
                    }
                }

                // Sort the elements of the submatrix to easily find the minimum difference
                std::sort(submatrix_elements.begin(), submatrix_elements.end());

                // Calculate the minimum absolute difference between distinct elements
                int min_diff = INT_MAX;
                bool all_same = true; // Flag to check if all elements are the same

                // Iterate through the sorted elements to find the minimum difference
                // We only need to check adjacent elements after sorting
                for (size_t i = 0; i < submatrix_elements.size() - 1; ++i) {
                    // If we find two different adjacent elements, the flag is set to false
                    if (submatrix_elements[i] != submatrix_elements[i + 1]) {
                        all_same = false;
                        // Calculate the difference and update min_diff if it's smaller
                        min_diff = std::min(min_diff, submatrix_elements[i + 1] - submatrix_elements[i]);
                    }
                }

                // If all elements in the submatrix were the same, the minimum difference is 0.
                // Otherwise, it's the min_diff we calculated.
                if (all_same) {
                    ans[r][c] = 0;
                } else {
                    ans[r][c] = min_diff;
                }
            }
        }

        return ans;
    }
};
```

File: solutions/2026-03-20/java/2026-03-20-minimum-absolute-difference-in-sliding-submatrix.java
```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Problem Summary:
 * Given an m x n integer matrix and an integer k, find the minimum absolute difference
 * between any two distinct values within each contiguous k x k submatrix.
 *
 * Link: https://leetcode.com/problems/minimum-absolute-difference-in-sliding-submatrix/
 *
 * Approach:
 * We can iterate through all possible top-left corners of the k x k submatrices.
 * For each submatrix, we extract all its elements into a temporary list, sort the list,
 * and then iterate through the sorted list to find the minimum difference between adjacent elements.
 * This minimum difference is the answer for that specific submatrix.
 *
 * To optimize the extraction and sorting for overlapping submatrices, we can use a sliding window approach.
 * However, given the small constraints (m, n <= 30), a direct approach of extracting, sorting, and finding the min diff
 * for each submatrix is sufficient and easier to implement.
 *
 * For each submatrix:
 * 1. Extract all k*k elements.
 * 2. Sort these k*k elements.
 * 3. Calculate the minimum difference between adjacent elements in the sorted list.
 *
 * Time Complexity:
 * Let m be the number of rows and n be the number of columns in the grid.
 * The number of possible k x k submatrices is (m - k + 1) * (n - k + 1).
 * For each submatrix, we extract k*k elements, sort them (O(k^2 log(k^2))), and find the minimum difference (O(k^2)).
 * The dominant part is sorting. So, the total time complexity is O((m-k+1)*(n-k+1) * k^2 log(k^2)).
 * Since m, n <= 30, k <= min(m, n), the maximum k^2 log(k^2) is around 30*30*log(30*30) which is manageable.
 *
 * Space Complexity:
 * For each submatrix, we create a temporary list to store k*k elements. This takes O(k^2) space.
 * The result matrix `ans` takes O((m-k+1)*(n-k+1)) space.
 * The overall space complexity is O(k^2 + (m-k+1)*(n-k+1)). Since k <= min(m, n), the dominant term is typically O(m*n).
 */
class Solution {
    public int[][] minimumAbsDifference(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;

        // The result matrix will have dimensions (m - k + 1) x (n - k + 1)
        int[][] ans = new int[m - k + 1][n - k + 1];

        // Iterate through all possible top-left corners of the k x k submatrices
        for (int i = 0; i <= m - k; i++) {
            for (int j = 0; j <= n - k; j++) {
                // For each submatrix, we need to find the minimum absolute difference.
                // We'll store all elements of the current submatrix in a list.
                List<Integer> submatrixElements = new ArrayList<>();

                // Extract all elements from the current k x k submatrix
                for (int row = i; row < i + k; row++) {
                    for (int col = j; col < j + k; col++) {
                        submatrixElements.add(grid[row][col]);
                    }
                }

                // Sort the elements of the submatrix to easily find the minimum difference
                Collections.sort(submatrixElements);

                // Initialize minimum difference to a large value
                int minDiff = Integer.MAX_VALUE;

                // If all elements are the same, the difference is 0.
                // This is implicitly handled if the list has only one unique element,
                // but we can also check explicitly if the size of unique elements is 1.
                // If list size is less than 2, minDiff remains MAX_VALUE, which will be 0 if there's only one element.
                if (submatrixElements.size() > 1) {
                    // Iterate through the sorted list and find the minimum difference between adjacent elements
                    for (int l = 0; l < submatrixElements.size() - 1; l++) {
                        minDiff = Math.min(minDiff, submatrixElements.get(l + 1) - submatrixElements.get(l));
                    }
                } else {
                    // If there's only one element (k=1 and submatrix has one element), the min diff is 0.
                    minDiff = 0;
                }
                
                // Store the calculated minimum difference in the result matrix
                ans[i][j] = minDiff;
            }
        }

        // Return the 2D array containing the minimum absolute differences for each submatrix
        return ans;
    }
}
```